File: Android/RTMSyncManager/src/main/java/io/agora/rtmsyncmanager/service/imp/AUIUserServiceImpl.kt
package io.agora.rtmsyncmanager.service.imp

import io.agora.rtmsyncmanager.model.AUIRoomContext
import io.agora.rtmsyncmanager.model.AUIUserInfo
import io.agora.rtmsyncmanager.service.IAUIUserService
import io.agora.rtmsyncmanager.service.callback.AUICallback
import io.agora.rtmsyncmanager.service.callback.AUIException
import io.agora.rtmsyncmanager.service.callback.AUIUserListCallback
import io.agora.rtmsyncmanager.service.rtm.AUIRtmException
import io.agora.rtmsyncmanager.service.rtm.AUIRtmManager
import io.agora.rtmsyncmanager.service.rtm.AUIRtmUserLeaveReason
import io.agora.rtmsyncmanager.service.rtm.AUIRtmUserRespObserver
import io.agora.rtmsyncmanager.utils.AUILogger
import io.agora.rtmsyncmanager.utils.GsonTools
import io.agora.rtmsyncmanager.utils.ObservableHelper

class AUIUserServiceImpl constructor(
    private val channelName: String,
    private val rtmManager: AUIRtmManager
) : IAUIUserService, AUIRtmUserRespObserver {

    private val TAG = "AUiUserServiceImpl"

    private var mUserList = mutableListOf<AUIUserInfo>()

    private val observableHelper =
        ObservableHelper<IAUIUserService.AUIUserRespObserver>()

    init {
        rtmManager.subscribeUser(this)
    }

    fun release() {
        observableHelper.unSubscribeAll()
        rtmManager.unsubscribeUser(this)
    }

    override fun registerRespObserver(observer: IAUIUserService.AUIUserRespObserver?) {
        observableHelper.subscribeEvent(observer)
        if(mUserList.isNotEmpty()){
            this.observableHelper.notifyEventHandlers {
                it.onRoomUserSnapshot(channelName, mUserList)
            }
        }
    }

    override fun unRegisterRespObserver(observer: IAUIUserService.AUIUserRespObserver?) {
        observableHelper.unSubscribeEvent(observer)
    }

    override fun getUserInfoList(
        roomId: String,
        callback: AUIUserListCallback?
    ) {
        rtmManager.whoNow(roomId) { error, userList ->
            if (error != null) {
                callback?.onResult(
                    AUIException(
                        error.code,
                        error.message
                    ), null)
                return@whoNow
            }
            if (userList == null) {
                callback?.onResult(
                    AUIException(
                        -1,
                        ""
                    ), null)
                return@whoNow
            }
            val users = mutableListOf<AUIUserInfo>()
            userList.forEach { userMap ->
                GsonTools.toBean(GsonTools.beanToString(userMap), AUIUserInfo::class.java)?.let {
                    users.add(it)
                }
            }
            callback?.onResult(null, users)
        }
    }

    override fun muteUserAudio(isMute: Boolean, callback: AUICallback?) {
        val currentUserId = roomContext.currentUserInfo.userId
        val user = mUserList.first { it.userId == currentUserId }
        user.muteAudio = isMute
        val map = GsonTools.beanToMap(user)
        rtmManager.setPresenceState(channelName, attr = map.mapValues { it.value.toString() }) { error ->
            if (error != null) {
                callback?.onResult(
                    AUIException(
                        error.code,
                        error.message
                    )
                )
            } else {
                this.observableHelper.notifyEventHandlers {
                    it.onUserAudioMute(currentUserId, isMute)
                }
                callback?.onResult(null)
            }
        }
    }

    override fun muteUserVideo(isMute: Boolean, callback: AUICallback?) {
        val currentUserId = roomContext.currentUserInfo.userId
        val user = mUserList.firstOrNull { it.userId == currentUserId }
        if (user == null) {
            callback?.onResult(
                AUIException(
                    -1,
                    "can't find current user from users"
                )
            )
            return
        }
        user.muteVideo = isMute
        val map = GsonTools.beanToMap(user) as Map<String, String>
        rtmManager.setPresenceState(channelName, attr = map) { error ->
            if (error != null) {
                callback?.onResult(
                    AUIException(
                        error.code,
                        error.message
                    )
                )
            } else {
                callback?.onResult(null)
            }
        }
    }

    override fun getUserInfo(userId: String): AUIUserInfo? {
        return mUserList.firstOrNull { it.userId == userId }
    }

    override fun getRoomContext(): AUIRoomContext { return AUIRoomContext.shared() }

    override fun getChannelName() = channelName

    /** AUiRtmUserProxyDelegate */
    override fun onUserSnapshotRecv(
        channelName: String,
        userId: String,
        userList: List<Map<String, Any>>
    ) {
        if (this.channelName != channelName) return
        val users = mutableListOf<AUIUserInfo>()
        userList.forEach { userMap ->
            GsonTools.toBean(GsonTools.beanToString(userMap), AUIUserInfo::class.java)?.let {
                users.add(it)
            }
        }
        mUserList = users
        this.observableHelper.notifyEventHandlers {
            it.onRoomUserSnapshot(channelName, mUserList)
        }
    }

    override fun onUserDidJoined(
        channelName: String,
        userId: String,
        userInfo: Map<String, Any>
    ) {
        if (this.channelName != channelName) return
        if (userInfo.size <= 1) return
        GsonTools.toBean(GsonTools.beanToString(userInfo), AUIUserInfo::class.java)?.let { info ->
            info.userId = userId
            mUserList.add(info)
            this.observableHelper.notifyEventHandlers {
                it.onRoomUserEnter(channelName, info)
            }
        }
    }

    override fun onUserDidLeaved(
        channelName: String,
        userId: String,
        userInfo: Map<String, Any>,
        reason: AUIRtmUserLeaveReason
    ) {
        if (this.channelName != channelName) return
        val index = mUserList.indexOfFirst{ it.userId == userId }
        val info = mUserList.removeAt(index)
        this.observableHelper.notifyEventHandlers {
            it.onRoomUserLeave(channelName, info, reason)
        }
    }

    override fun onUserDidUpdated(
        channelName: String,
        userId: String,
        userInfo: Map<String, Any>
    ) {
        if (this.channelName != channelName) return
        if (userInfo.isEmpty()) {
            return
        }
        GsonTools.toBean(GsonTools.beanToString(userInfo), AUIUserInfo::class.java)?.let { info ->
            val index = mUserList.indexOfFirst{ it.userId == info.userId }
            if (index == -1) { // 不存在该用户
                mUserList.add(info)
                this.observableHelper.notifyEventHandlers {
                    it.onRoomUserEnter(channelName, info)
                }
                return
            } else {
                val oldInfo = mUserList[index]
                mUserList[index] = info
                // 单独更新语音被禁用回调
                if (oldInfo.muteAudio != info.muteAudio) {
                    this.observableHelper.notifyEventHandlers {
                        it.onUserAudioMute(info.userId, info.muteAudio)
                    }
                }
            }
            this.observableHelper.notifyEventHandlers {
                it.onRoomUserUpdate(channelName, info)
            }
        }
    }

    fun setUserPayload(payload: String) {
        val roomId = channelName
        val userId = AUIRoomContext.shared().currentUserInfo.userId
        AUILogger.logger().d(TAG, "setPayload[$channelName] : $payload")
        val userAttr = mapOf("customPayload" to payload)
        rtmManager.setPresenceState(channelName = roomId, attr = userAttr) { error ->
            if (error != null) {
                AUILogger.logger().d(TAG, "setPayload[$roomId] fail: ${error.localizedMessage}")
                //TODO: retry
                return@setPresenceState
            }

            // rtm不会返回自己更新的数据，需要手动处理
            onUserDidUpdated(channelName = roomId, userId = userId, userInfo = userAttr)
        }
    }


    fun setUserAttr(completion: (AUIRtmException?) -> Unit) {
        val roomId = channelName
        val userId = AUIRoomContext.shared().currentUserInfo.userId
        val userInfo = mUserList.firstOrNull { it.userId == userId } ?: AUIUserInfo()
        userInfo.userId = AUIRoomContext.shared().currentUserInfo.userId
        userInfo.userName = AUIRoomContext.shared().currentUserInfo.userName
        userInfo.userAvatar = AUIRoomContext.shared().currentUserInfo.userAvatar

        val userAttr = GsonTools.beanToMap(userInfo)
        AUILogger.logger().d(TAG, "setupUserAttr: $roomId : $userAttr")
        rtmManager.setPresenceState(roomId, attr = userAttr) { error ->
            if (error != null) {
                AUILogger.logger().d(TAG, "setupUserAttr: $roomId fail: ${error.reason}")
            } else {
                //rtm不会返回自己更新的数据，需要手动处理
                onUserDidUpdated(roomId, userId, userAttr)
            }
        }
    }
}

File: iOS/AgoraEntScenarios/Common/API/RTMSyncManager/RTMSyncManager/Classes/Service/Impl/AUIUserServiceImpl.swift
//  AUIUserServiceImpl.swift
//  AUIKit
//
//  Created by wushengtao on 2023/4/7.
//

import Foundation
import AgoraRtmKit
import YYModel

@objc open class AUIUserServiceImpl: NSObject {
    public private(set) var userList: [AUIUserInfo] = []
    private var respDelegates = NSHashTable<AUIUserRespDelegate>.weakObjects()
    private var channelName: String
    private let rtmManager: AUIRtmManager
    
    deinit {
        aui_info("deinit AUIUserServiceImpl[\(channelName)]", tag: "AUIUserServiceImpl")
        rtmManager.unsubscribeUser(channelName: channelName, delegate: self)
    }
    
    public init(channelName: String, rtmManager: AUIRtmManager) {
        self.rtmManager = rtmManager
        self.channelName = channelName
        super.init()
        aui_info("init AUIUserServiceImpl[\(channelName)]", tag: "AUIUserServiceImpl")
        self.rtmManager.subscribeUser(channelName: channelName, delegate: self)
        //rtm2.2 支持预设置，在subscribe成功之后会更新
//        _setupUserAttr(roomId: channelName) { _ in
//            //TODO: retry
//        }
    }
}

//MARK: AUIRtmUserProxyDelegate
extension AUIUserServiceImpl: AUIRtmUserProxyDelegate {
    public func onCurrentUserJoined(channelName: String) {
        guard channelName == self.channelName else {return}
        aui_info("onCurrentUserJoined[\(channelName)]", tag: "AUIUserServiceImpl")
        let user = AUIUserInfo(thumbUser: AUIRoomContext.shared.currentUserInfo)
        self.respDelegates.allObjects.forEach { obj in
            obj.onRoomUserEnter(roomId: channelName, userInfo: user)
        }
    }
    
    public func onUserDidUpdated(channelName: String, userId: String, userInfo: [String : Any]) {
        aui_info("onUserDidUpdated[\(channelName)]: \(userId) \(userInfo)", tag: "AUIUserServiceImpl")
        guard let user = AUIUserInfo.yy_model(withJSON: userInfo) else { return }
        guard let idx = self.userList.firstIndex(where: {$0.userId == userId}) else {
            onUserDidJoined(channelName: channelName, userId: userId, userInfo: userInfo)
            return
        }
        
        user.userId = userId
        if let oldUser = self.userList.first(where: {$0.userId == userId}) {
            if oldUser.muteAudio != user.muteAudio {
                oldUser.muteAudio = user.muteAudio
                self.respDelegates.allObjects.forEach { obj in
                    obj.onUserAudioMute(userId: userId, mute: user.muteAudio)
                }
            }
            
            if oldUser.muteVideo != user.muteVideo {
                oldUser.muteVideo = user.muteVideo
                self.respDelegates.allObjects.forEach { obj in
                    obj.onUserVideoMute(userId: userId, mute: user.muteVideo)
                }
            }
        }
        
        self.userList.replaceSubrange(idx...idx, with: [user])
        self.respDelegates.allObjects.forEach { obj in
            obj.onRoomUserUpdate(roomId: channelName, userInfo: user)
        }
    }
    
    public func onUserSnapshotRecv(channelName: String, userId: String, userList: [[String : Any]]) {
        aui_info("onUserSnapshotRecv[\(channelName)]: \(userId) \(userList)", tag: "AUIUserServiceImpl")
        guard let users = NSArray.yy_modelArray(with: AUIUserInfo.self, json: userList) as? [AUIUserInfo] else {
            assert(false, "onUserSnapshotRecv recv fail")
            return
        }
        self.respDelegates.allObjects.forEach { obj in
            self.userList = users
            obj.onRoomUserSnapshot(roomId: channelName, userList: users)
        }
//        _setupUserAttr() { _ in
//            //TODO: retry
//        }
    }
    
    public func onUserDidJoined(channelName: String, userId: String, userInfo: [String : Any]) {
        aui_info("onUserDidJoined[\(channelName)]: \(userId) \(userInfo)", tag: "AUIUserServiceImpl")
        if userInfo.count <= 1 {
            return
        }
        let user = AUIUserInfo.yy_model(withJSON: userInfo)!
        user.userId = userId
        self.userList.append(user)
        self.respDelegates.allObjects.forEach { obj in
            obj.onRoomUserEnter(roomId: channelName, userInfo: user)
        }
    }
    
    public func onUserDidLeaved(channelName: String, userId: String, userInfo: [String : Any], reason: AUIRtmUserLeaveReason) {
        aui_info("onUserDidLeaved[\(channelName)]: \(userId) \(userInfo)", tag: "AUIUserServiceImpl")
        let user = userList.filter({$0.userId == userId}).first ?? AUIUserInfo.yy_model(withJSON: userInfo)!
        self.userList = userList.filter({$0.userId != userId})
        self.respDelegates.allObjects.forEach { obj in
            obj.onRoomUserLeave(roomId: channelName, userInfo: user, reason: reason)
        }
    }
    
    public func onUserBeKicked(channelName: String, userId: String, userInfo: [String : Any]) {
        aui_info("onUserBeKicked[\(channelName)]: \(userId) \(userInfo)", tag: "AUIUserServiceImpl")
        let user = userList.filter({$0.userId == userId}).first ?? AUIUserInfo.yy_model(withJSON: userInfo)!
        self.userList = userList.filter({$0.userId != userId})
        self.respDelegates.allObjects.forEach { obj in
            obj.onUserBeKicked(roomId: channelName, userId: user.userId)
        }
    }
}

extension AUIUserServiceImpl: AUIUserServiceDelegate {
    
    public func getRoomContext() -> AUIRoomContext {
        return AUIRoomContext.shared
    }
    
    public func getChannelName() -> String {
        return channelName
    }
    
    public func bindRespDelegate(delegate: AUIUserRespDelegate) {
        aui_info("bindRespDelegate[\(channelName)] \(delegate)", tag: "AUIUserServiceImpl")
        respDelegates.add(delegate)
    }
    
    public func unbindRespDelegate(delegate: AUIUserRespDelegate) {
        aui_info("unbindRespDelegate[\(channelName)] \(delegate)", tag: "AUIUserServiceImpl")
        respDelegates.remove(delegate)
    }
    
    public func getUserInfoList(roomId: String, callback:@escaping AUIUserListCallback) {
        aui_info("getUserInfoList[\(channelName)]", tag: "AUIUserServiceImpl")
        self.rtmManager.whoNow(channelName: roomId) { error, userList in
            if let error = error {
                callback(error as NSError, nil)
                return
            }
            
            var users = [AUIUserInfo]()
            userList?.forEach { attr in
                let user = AUIUserInfo.yy_model(withJSON: attr)!
                users.append(user)
            }
            self.userList = users
            callback(nil, users)
        }
    }
    
//    public func getUserInfo(by userId: String) -> AUIUserThumbnailInfo? {
//        return self.userList.filter {$0.userId == userId}.first
//    }
    
    public func muteUserAudio(isMute: Bool, callback: @escaping AUICallback) {
        aui_info("muteUserAudio[\(channelName)]: \(isMute)", tag: "AUIUserServiceImpl")
        let currentUserId = getRoomContext().currentUserInfo.userId
        let currentUser = userList.first(where: {$0.userId == currentUserId})
        currentUser?.muteAudio = isMute
        let userAttr = currentUser?.yy_modelToJSONObject() as? [String: Any] ?? [:]
//        print("muteUserAudio user:  \(userDic)")
        self.rtmManager.setPresenceState(channelName: channelName, attr: userAttr) {[weak self] error in
            guard let self = self else {return}
            if let error = error {
                callback(error as NSError)
                return
            }
            
            callback(nil)
            
            //自己状态不会更新，在这里手动回调
            self.respDelegates.allObjects.forEach { obj in
                obj.onUserAudioMute(userId: currentUserId, mute: isMute)
            }
        }
    }
    
    public func muteUserVideo(isMute: Bool, callback: @escaping AUICallback) {
        aui_info("muteUserVideo[\(channelName)]: \(isMute)", tag: "AUIUserServiceImpl")
        let currentUserId = getRoomContext().currentUserInfo.userId
        let currentUser = userList.first(where: {$0.userId == currentUserId})
        currentUser?.muteVideo = isMute
        let userAttr = currentUser?.yy_modelToJSONObject() as? [String: Any] ?? [:]
//        print("muteUserAudio user:  \(userDic)")
        self.rtmManager.setPresenceState(channelName: channelName, attr: userAttr) {[weak self] error in
            guard let self = self else {return}
            if let error = error {
                callback(error as NSError)
                return
            }
            
            callback(nil)
            
            //自己状态不会更新，在这里手动回调
            self.respDelegates.allObjects.forEach { obj in
                obj.onUserVideoMute(userId: currentUserId, mute: isMute)
            }
        }
    }
    
    public func kickUser(roomId: String, userId: String, callback: @escaping AUICallback) {
//        let model = AUIKickUserReqModel()
//        model.roomId = roomId
//        model.uid = Int(userId) ?? 0
//        model.operatorId = getRoomContext().currentUserInfo.userId
//        model.request { error, obj in
//            callback(error as? NSError)
//        }
    }
}

extension AUIUserServiceImpl {
    func setUserPayload(payload: String) {
        let roomId = channelName
        let userId = AUIRoomContext.shared.currentUserInfo.userId
        aui_info("setPayload[\(channelName)] : \(payload)", tag: "AUIUserServiceImpl")
        let userAttr = ["customPayload": payload]
        self.rtmManager.setPresenceState(channelName: roomId, attr: userAttr) { error in
            if let error = error {
                aui_info("setPayload[\(roomId)] fail: \(error.localizedDescription)", tag: "AUIUserServiceImpl")
                //TODO: retry
                return
            }
            
            //rtm不会返回自己更新的数据，需要手动处理
            self.onUserDidUpdated(channelName: roomId, userId: userId, userInfo: userAttr)
        }
    }
    
    //设置用户属性到presence
    func setUserAttr(completion: ((Error?) -> ())?) {
        let roomId = channelName
        let userId = AUIRoomContext.shared.currentUserInfo.userId
        let userInfo = self.userList.filter({$0.userId == userId}).first ?? AUIUserInfo()
        userInfo.userId = AUIRoomContext.shared.currentUserInfo.userId
        userInfo.userName = AUIRoomContext.shared.currentUserInfo.userName
        userInfo.userAvatar = AUIRoomContext.shared.currentUserInfo.userAvatar
        
        let userAttr = userInfo.yy_modelToJSONObject() as? [String: Any] ?? [:]
        aui_info("setUserAttr[\(channelName)] : \(userAttr)", tag: "AUIUserServiceImpl")
        self.rtmManager.setPresenceState(channelName: roomId, attr: userAttr) { error in
            defer {
                completion?(error)
            }
            if let error = error {
                aui_info("setupUserAttr[\(roomId)] fail: \(error.localizedDescription)", tag: "AUIUserServiceImpl")
                //TODO: retry
                return
            }
            
            //rtm不会返回自己更新的数据，需要手动处理
            self.onUserDidUpdated(channelName: roomId, userId: userId, userInfo: userAttr)
        }
    }
}

